[FLINK-25171] Validation of duplicate fields in ddl sql

jelly-1203 opened the pull request:
## What is the purpose of the change

The purpose of this pull Request is to add validation to the derived table that the fields are duplicated


## Brief change log

-  Added verification of whether the derived table fields are duplicated


## Verifying this change

- Added tests to verify the exception is thrown in the correct scenarios

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

wenlong88 commented:
Thanks for the contribution, @jelly-1203 , you may need to add some test on the change. 
By the way, you can do some clean up on description of pr. The template is a guidance to help you fill in information which is important for reviewers.

jelly-1203 commented:
> Thanks for the contribution, @jelly-1203 , you may need to add some test on the change. By the way, you can do some clean up on description of pr. The template is a guidance to help you fill in information which is important for reviewers.

@wenlong88 Thanks for your comments, I will add tests for this change later and tidy up the description of this pr

jelly-1203 commented:
> Thanks for the contribution, @jelly-1203 , you may need to add some test on the change. By the way, you can do some clean up on description of pr. The template is a guidance to help you fill in information which is important for reviewers.

@wenlong88 Hello, I have added test for this change and modified the description of this pr, please help to review it again

xuyangzhong commented:
Hi, thanks for your contribution, @jelly-1203 . It seems that you have a bad code style in your new code. You can see it in CI report above. You can follow this doc to set the code style to automate it in IDEA: https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting

jelly-1203 commented:
> Hi, thanks for your contribution, @jelly-1203 . It seems that you have a bad code style in your new code. You can see it in CI report above. You can follow this doc to set the code style to automate it in IDEA: https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting

@xuyangzhong  Thanks for your comment. I have fixed and verified the code style. Please help to review it again

jelly-1203 commented:
> Hi, thanks for your contribution, @jelly-1203 . It seems that you have a bad code style in your new code. You can see it in CI report above. You can follow this doc to set the code style to automate it in IDEA: https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting

Hi, @xuyangzhong  thanks for your  comments, I have to make the changes to your code style, compilation is still not passed, by looking at the error message, found in the org.apache.flink.table.planner.plan.stream.sql.UnionTest, An SQL statement with repeated fields was executed, so the compilation failed. Please kindly ask whether I need to mention an issue and modify the SQL?
Error code:
    util.tableEnv.executeSql(
      s"""
         |CREATE TABLE t1 (
         |  id int,
         |  ts bigint,
         |  name string,
         |  timestamp_col timestamp(3),
         |  val bigint,
         |  name varchar(32),
         |  timestamp_ltz_col as TO_TIMESTAMP_LTZ(ts, 3),
         |  watermark for timestamp_col as timestamp_col
         |) WITH (
         |  'connector' = 'values',
         |  'bounded' = 'false'
         |)
       """.stripMargin)

xuyangzhong commented:
Hi, @jelly-1203. You can rename one of the "name" for convenience. You needn't to mention your issue here because the GIT can record this change. If someone confuses this modifier, he can find your issue. So just modify it directly and code review will verify this correctness about this modifier.

jelly-1203 commented:
> Hi, @jelly-1203. You can rename one of the "name" for convenience. You needn't to mention your issue here because the GIT can record this change. If someone confuses this modifier, he can find your issue. So just modify it directly and code review will verify this correctness about this modifier. BTW, you may also need to modify the generated plan in UnionTest.xml.

Hi, @xuyangzhong, Thanks for your comment, I will modify it and note the match to the logical plan in uniontest.xml

xuyangzhong commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/UnionTest.scala:
```diff
@@ -40,10 +40,9 @@ class UnionTest extends TableTestBase {
          |  id int,
          |  ts bigint,
-         |  name string,
+         |  name varchar(32),
```
If you want to delete one of them, I think it's better to keep the one that has the same type with other tables.

xuyangzhong commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java:
```diff
@@ -494,7 +494,13 @@ private void collectPhysicalFieldsTypes(List<SqlNode> derivedColumns) {
+                    if (oldType != null) {
+                        throw new ValidationException(
+                                String.format(
+                                        "A column named '%s' already exists in the derived table.",
```
The exception message is confusing where the derived table is, because users maybe only use "create table ..." instead of "create table ... like ...". But the previous one can also cause this bug. IMO, you can delete the word 'derived'.

xuyangzhong reviewed (commented):
I have left some comments following.

jelly-1203 commented:
> I have left some comments following.

Hi,@xuyangzhong  Thanks for your advice. I will make adjustments as soon as possible

wenlong88 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java:
```diff
@@ -494,7 +494,13 @@ private void collectPhysicalFieldsTypes(List<SqlNode> derivedColumns) {
                     RelDataType relType = type.deriveType(sqlValidator, nullable);
                     // add field name and field type to physical field list
-                    physicalFieldNamesToTypes.put(name, relType);
+                    RelDataType oldType = physicalFieldNamesToTypes.put(name, relType);
```
I think it is not enough to add check here, when there is name conflicts between computedColumn or metadata Column, the check here would not work well. you can try to add validation in appendDerivedColumns.

jelly-1203 replied:
> I think it is not enough to add check here, when there is name conflicts between computedColumn or metadata Column, the check here would not work well. you can try to add validation in appendDerivedColumns.

Hi, @wenlong88 Thanks for your advice. I think it is very meaningful. I will verify it first and make adjustments in time if there is any problem

jelly-1203 replied:
> I think it is not enough to add check here, when there is name conflicts between computedColumn or metadata Column, the check here would not work well. you can try to add validation in appendDerivedColumns.

Hi, @wenlong88  Thanks for your advice, which is of great help to me. I found several problems during the testing process
1. Duplicate columns are overwritten when the compute column is first and the regular column is last
2. If the metadata column is first, use the compute column in the middle and the regular column in the rear. If the metadata column name is the same as the regular column name, use the compute column in the middle and the regular column name. In the generated comput column process, namely in accessibleFieldNamesToTypes. PutAll methods covered in repeated fields.

I have adjusted the code accordingly and added test for adjustment. Please help to review it

wenlong88 reviewed (changes requested):
Thanks for the contribution, I left a comment. I think it is not a good place where the validation added currently.

wenlong88 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java:
```diff
@@ -426,6 +433,14 @@ private void appendDerivedColumns(
+                    if (!result.isEmpty()) {
+                        throw new ValidationException(
+                                "A field name conflict exists between a field of the regular type and a field of the Metadata type.");
+                    }
```
may be we can just check duplication when put the new Column to the columns, at the end of this function?

jelly-1203 replied:
> may be we can just check duplication when put the new Column to the columns, at the end of this function?

hi @wenlong88 
Thanks for your review and comments, I think we can adjust the determination of whether a SqlRegularColumn type field is duplicated with other types of fields to check if a new column is duplicated when put into the column at the end of this function.
But accessibleFieldNamesToTypes heavy logic to this place, I think it is necessary, because, when creating computecolcolumn if metadata columns in the former, repetitive physical column in the bottom, In the use of accessibleFieldNamesToTypes putAll, metadata list field covers physical column, if the user want, according to the calculated regularColumn you don't need to do so.

jelly-1203 replied:
> may be we can just check duplication when put the new Column to the columns, at the end of this function?

hi @wenlong88 
Thanks for your review and comments, I think we can adjust the determination of whether a SqlRegularColumn type field is duplicated with other types of fields to check if a new column is duplicated when put into the column at the end of this function.
But accessibleFieldNamesToTypes heavy logic to this place, I think it is necessary, because, when creating computecolcolumn if metadata columns in the former, repetitive physical column in the bottom, In the use of accessibleFieldNamesToTypes putAll, metadata list field covers physical column, if the user want, according to the calculated regularColumn you don't need to do so.

jelly-1203 replied:
> may be we can just check duplication when put the new Column to the columns, at the end of this function?

hi,@wenlong88 Do you think my view is correct?

wenlong88 replied:
hi, @jelly-1203 thanks for the update and  analysis, I agree that it not  possible to add unified check at the end. but I still think that the check here can be improved a bit:
1. it seems that the check here is not relevant to the new added ComputingColumn, if you want to check the duplication, it is better to add it when updating metadataFieldNamesToTypes.
2. according to current implementation, regular columns have top priority(we collect collectPhysicalFieldsTypes at the beginning), so we may also need to check if there is duplicated name in physicalFieldNamesToTypes when trying to a metadata column or computed column. If we add such check, the check in 1 is not necessary any more.

what do you think?

jelly-1203 replied:
> hi, @jelly-1203 thanks for the update and analysis, I agree that it not possible to add unified check at the end. but I still think that the check here can be improved a bit:
> 
> 1. it seems that the check here is not relevant to the new added ComputingColumn, if you want to check the duplication, it is better to add it when updating metadataFieldNamesToTypes.
> 2. according to current implementation, regular columns have top priority(we collect collectPhysicalFieldsTypes at the beginning), so we may also need to check if there is duplicated name in physicalFieldNamesToTypes when trying to a metadata column or computed column. If we add such check, the check in 1 is not necessary any more.
> 
> what do you think?

hi, @wenlong88 I think your suggestion is reasonable, I will try to modify it

jelly-1203 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java:
```diff
@@ -426,6 +433,14 @@ private void appendDerivedColumns(
+                    if (!result.isEmpty()) {
+                        throw new ValidationException(
+                                "A field name conflict exists between a field of the regular type and a field of the Metadata type.");
+                    }
```
Hi @wenlong88 
Thanks for your review and comments, I think we can adjust the determination of whether a SqlRegularColumn type field is duplicated with other types of fields to check if a new column is duplicated when put into the column at the end of this function.
But accessibleFieldNamesToTypes heavy logic to this place, I think it is necessary, because, when creating computecolcolumn if metadata columns in the former, repetitive physical column in the bottom, In the use of accessibleFieldNamesToTypes putAll, metadata list field covers physical column, if the user want, according to the calculated regularColumn you don't need to do so.

jelly-1203 replied:
Hi @wenlong88 
Thanks for your review and comments, I think we can adjust the determination of whether a SqlRegularColumn type field is duplicated with other types of fields to check if a new column is duplicated when put into the column at the end of this function.
But accessibleFieldNamesToTypes heavy logic to this place, I think it is necessary, because, when creating computecolcolumn if metadata columns in the former, repetitive physical column in the bottom, In the use of accessibleFieldNamesToTypes putAll, metadata list field covers physical column, if the user want, according to the calculated regularColumn you don't need to do so.

jelly-1203 commented:
Hi, @wenlong88 Please help to review it and thank you for your suggestions

wenlong88 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/MergeTableLikeUtilTest.java:
```diff
@@ -126,6 +126,72 @@ public void mergeWithIncludeFailsOnDuplicateColumn() {
+                        regularColumn("four", DataTypes.STRING()));
+
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage(
```
in this case, I think it would be better to throw with information that there are duplicate column with name 'two' in metadata column and regular column?

jelly-1203 replied:
> Contributor

Ok, I'll return this error message more explicitly

jelly-1203 commented:
Hi, @wenlong88 I have adjusted the position of the verification logic and made the error message return more clear. Could you please help to review it and see what needs to be improved? Thank you

jelly-1203 commented:
Hi, @wenlong88 Could you please review it and see what needs to be improved

wenlong88 commented:
LGTM，cc @godfreyhe  to do the final check

jelly-1203 commented:
Hi, @godfreyhe 
Please have time to do the final check. If there is any deficiency, I will continue to improve

jelly-1203 commented:
Hi,@wenlong88  could you please ping @godfreyhe again? There is no progress on this issue

jelly-1203 commented:
anyone?

wenlong88 commented:
@jelly-1203  thanks for the following up, I would ping @godfreyhe  offline to follow the pr.

godfreyhe reviewed (approved):
Sorry for the late response, LGTM, I will merge it
